drud/ddev
# Apache for ddev, fixes #628

## The Problem/Issue/Bug:

Lots of users want Apache, see OP #628, mostly for htaccess types of situations.

We expect this to be an "experimental support" situation; full tests may not be in place when we release, and we can then learn from people's experience.

## How this PR Solves The Problem:

Provides the ability to configure the webserver type in .ddev/config.yaml, you can use 

```
webserver_type: apache-fpm
webserver_type: apache-cgi
webserver_type: nginx-fpm (the default)
```

TODO:
- [x] Custom configuration for apache (I expect we'll spin this off into some other issue)
- [x] mod_rewrite (for drupal rewrites) doesn't seem to be working yet.
- [x] ddev logs isn't showing the right stuff for apache
- [x] docroot needs to be substituted/updated in apache config.
- [x] apache-cgi not yet working
- [ ] Tests of course. Complete tests for all permutations will have to be deferred to improvements of circle tests and parallelization
- [ ] Spin off issue for full testing of all apache2 permutations (parallel!)
- [x] Docs (config override, etc.)

This PR is currently in the "take it for a spin and report your results but don't take it too seriously" state. 

I think we might be able to enable this as "experimental" in the v1.2.0 release.

## Manual Testing Instructions:

* Make sure you have the latest ddev-webserver with `docker pull drud/ddev-webserver:20180815_apache`
* `ddev config` and edit your .ddev/config.yaml to use the webserver_type you prefer.
`ddev start`
* Experiment. On Drupal, you can visit admin/reports/status to see the webserver, etc.
* Note that as of 2018-08-08, this doesn't work with a docroot outside the project root yet.


## Automated Testing Overview:

This has no tests yet.

## Related Issue Link(s):

OP #628 

## Release/Deployment notes:
